api/src/Repository/DepartmentRepository.php

```diff
@@ -0,0 +1,63 @@
+<?php declare(strict_types=1);
+
+namespace App\Repository;
+
+use Atlas\Query\Select;
+
+class DepartmentRepository
+{
+
+    protected $db;
+
+
+    public function __construct($db)
+    {
+        $this->db = $db;
+
+    }
+
+
+    public function findAll()
+    {
+        $select = Select::new($this->db);
+        $select->columns(
+            'depts.DepartmentID',
+            'depts.Name',
+            'parentDepts.DepartmentID as ParentID',
+            'parentDepts.Name as ParentName',
+            'fallbackDepts.DepartmentID as FallbackID',
+            'fallbackDepts.Name as FallbackName'
+        )
+            ->columns(
+                $select->subSelect()
+                    ->columns('COUNT(DepartmentID)')
+                    ->from('Departments as childDepts')
+                    ->where('childDepts.ParentDepartmentID = depts.DepartmentID')
+                    ->andWhere('Name != "Historical Placeholder"')
+                    ->as('ChildCount')
+                    ->getStatement()
+            )
+            ->columns(
+                $select->subSelect()
+                    ->columns('GROUP_CONCAT(Email)')
+                    ->from('EMails')
+                    ->where('DepartmentID = depts.DepartmentID')
+                    ->as('Email')
+                    ->getStatement()
+            )
+            ->from('Departments as depts')
+            ->join('LEFT', 'Departments as parentDepts', 'parentDepts.DepartmentID = depts.ParentDepartmentID')
+            ->join('LEFT', 'Departments as fallbackDepts', 'fallbackDepts.DepartmentID = depts.FallbackID');
+
+        $historicalPlaceholders = $select->subSelect()->columns('DepartmentID')->from('Departments')->where('Name = "Historical Placeholder"');
+
+        $select->where('depts.DepartmentID NOT IN ', $historicalPlaceholders)
+            ->where('depts.ParentDepartmentID NOT IN ', $historicalPlaceholders);
+
+        return $select->fetchAll();
```

Comment:
So there is a pretty powerful system I build around the `columnsToAttributes` member.  I can understand wanting it pull it out into a `BaseRepository` or something out of the overloaded `BaseController`. I also notice it is not being used in the current Department code, which is disappointing, maybe because the merging of all the tables is being complicated, but it will still work, you can specified fields like `'<table>.<field>' => '<output>',` so `depts.DepartmentID =-> 'id',` would work and I feel is the better way to go.

Look at `Member/BaseMember.php` and see the definition for `protected static $columnsToAttributes` 
Then look at `Member/FindMember.php` and you can see `->columns(...BaseMember::selectMapping())` which programmatically builds all the Database field to resource value mapping. 

I see that a lot of that mapping has moved into the `service` below but it feels like there should be a cleaner way to express it, maybe still using a method like `columnsToAttributes`?  The reason SO many of the DB fields are different that the member fields is because the DB fields where inherited and do not fit a clean RESTful structure as I understood it.

Reply:
> So there is a pretty powerful system I build around the `columnsToAttributes` member. I can understand wanting it pull it out into a `BaseRepository` or something out of the overloaded `BaseController`.

I think `columnsToAttributes` could be useful in the future, but as we are implementing just a single repository here it did not feel necessary to make that abstraction at this time. While we can anticipate that there will be things in the future to work through, I think there are better options than to rely on another `Base` class and inheritance when we really want composition.

Additionally, it is not typical that the repository layer is doing any of this kind of mapping. That is typically reserved for the service layer in this pattern. The repository is typically responsible for making queries to the database and nothing more.

Perhaps another approach would be to make entity classes as a middle ground (which are also not using inheritance). In the constructor you can have it accept a row and it manages the mapping from SQL row to JSON on its own. We could also leverage an implementation for `equals` if we needed to do so, and if really needed, maybe some getters on the attributes.

I'd like to avoid adding complexity to our SQL queries because we should be able to pretty quickly put together what call is being made to the database. While `columnsToAttributes` wouldn't necessarily detract from this, I think there is a time and place where we could better implement this because it's very likely that we will want repository calls that _are not_ going to produce API output.

We also don't even have to map from Database column names to API friendly names (because I've seen `keyLikeThis`, `key_like_this` and `KeyLikeThis` come back from APIs. However it might be nice to hide the implementation (a little) or else we risk tightly coupling our UI code. I would make a strong case that this implementation should be hidden in the service layer, or with entities that represent the database objects, similar to how you'd see for an ORM library.

Reply:
> > So there is a pretty powerful system I build around the `columnsToAttributes` member. I can understand wanting it pull it out into a `BaseRepository` or something out of the overloaded `BaseController`.
> 
> I think `columnsToAttributes` could be useful in the future, but as we are implementing just a single repository here it did not feel necessary to make that abstraction at this time. While we can anticipate that there will be things in the future to work through, I think there are better options than to rely on another `Base` class and inheritance when we really want composition.
> 
> Additionally, it is not typical that the repository layer is doing any of this kind of mapping. That is typically reserved for the service layer in this pattern. The repository is typically responsible for making queries to the database and nothing more.
> 
> Perhaps another approach would be to make entity classes as a middle ground (which are also not using inheritance). In the constructor you can have it accept a row and it manages the mapping from SQL row to JSON on its own. We could also leverage an implementation for `equals` if we needed to do so, and if really needed, maybe some getters on the attributes.
> 
> I'd like to avoid adding complexity to our SQL queries because we should be able to pretty quickly put together what call is being made to the database. While `columnsToAttributes` wouldn't necessarily detract from this, I think there is a time and place where we could better implement this because it's very likely that we will want repository calls that _are not_ going to produce API output.
> 
> We also don't even have to map from Database column names to API friendly names (because I've seen `keyLikeThis`, `key_like_this` and `KeyLikeThis` come back from APIs. However it might be nice to hide the implementation (a little) or else we risk tightly coupling our UI code. I would make a strong case that this implementation should be hidden in the service layer, or with entities that represent the database objects, similar to how you'd see for an ORM library.

Yes I would like to "hide the implementation". 
I can see the argument for the mapping changing location and no longer having it part of the SQL query. I just would like to see a clean generalized system for it instead of each mapping layer having to have a messy large `output[X] = value[Y]` block.

Reply:
I'm in agreement, having a clean system for managing this is something we can work toward. I think I see a way we can incorporate this without having to use the functions that are directly on `BaseController` by extracting those into their own function. We might even be able to break them out and then let `BaseController` call them.

What can we do to handle nested objects? For example if we want to take `ParentID` and `ParentName` and put them in an output that is like so: `{ "parent": { "id": "123", "name": "The Name" } }`?

Reply:
> I'm in agreement, having a clean system for managing this is something we can work toward. I think I see a way we can incorporate this without having to use the functions that are directly on `BaseController` by extracting those into their own function. We might even be able to break them out and then let `BaseController` call them.
> 

I do not want to be taking steps back to take steps forward here.  Instead I would like sideways but non-regressive steps. 

> What can we do to handle nested objects? For example if we want to take `ParentID` and `ParentName` and put them in an output that is like so: `{ "parent": { "id": "123", "name": "The Name" } }`?

This is a bit of an extreme example.  But live..
When pulling a staff list with `api/department/2/staff` we get a response that is not unlike this:

```
{
    "type": "staff_list",
    "data": [
        {
            "type": "staff_entry",
            "id": "1022",
            "member": {
                "type": "member",
                "id": "1000",
                "legal_first_name": "Odin",
                "legal_last_name": "Allfather",
                "middle_name": null,
                "suffix": null,
                "email": "allfather@oneeye.com",
                "email2": null,
                "email3": null,
                "phone": null,
                "phone2": null,
                "address_line1": null,
                "address_line2": null,
                "city": null,
                "state": null,
                "zip_code": null,
                "zip_plus4": null,
                "country": null,
                "province": null,
                "preferred_first_name": null,
                "preferred_last_name": null,
                "deceased": null,
                "do_not_contact": null,
                "email_optout": null,
                "birthdate": null,
                "gender": "Allfather",
                "concom_display_phone": null,
                "dependent_on": null,
                "pronouns": null,
                "first_name": "Odin",
                "last_name": "Allfather",
                "duplicates": null
            },
            "department": {
                "id": "2",
                "parent": null,
                "name": "Activities",
                "fallback": {
                    "id": "8",
                    "parent": null,
                    "name": "Systems",
                    "fallback": {
                        "id": "7",
                        "parent": null,
                        "name": "Productions",
                        "fallback": {
                            "id": "6",
                            "parent": null,
                            "name": "Hospitality",
                            "fallback": {
                                "id": "5",
                                "parent": null,
                                "name": "Facilities",
                                "fallback": {
                                    "id": "4",
                                    "parent": null,
                                    "name": "External Relations and Communications",
                                    "fallback": {
                                        "id": "3",
                                        "parent": null,
                                        "name": "Administration",
                                        "fallback": "2",
                                        "child_count": "8",
                                        "email": [],
                                        "type": "department"
                                    },
                                    "child_count": "9",
                                    "email": [],
                                    "type": "department"
                                },
                                "child_count": "10",
                                "email": [],
                                "type": "department"
                            },
                            "child_count": "9",
                            "email": [],
                            "type": "department"
                        },
                        "child_count": "9",
                        "email": [],
                        "type": "department"
                    },
                    "child_count": "2",
                    "email": [],
                    "type": "department"
                },
                "child_count": "9",
                "email": [],
                "type": "department"
            },
            "positionId": "1",
            "note": "",
            "position": "Head"
        }
    ]
}
```

This is because the `staff_list` resource is a list of `staff_entry` and each `staff_entry` has a `member` field which expands into a `member` resource and a `department` field that expands into a `department` resource and so on.  This accordion of inclusions ensures that all the data that could be needed is present.  This is a bit of a mess because with the fallbacks we end up getting the whole fallback structure as well as other data we do not need.  This is likely part of the pain is that all these resources are getting queried over and over.  If we have a lot of people in this department then each `staff_entry` is going to have a `department` which is going to expand its fallback structure.

I think using a `service`, did i pick the right layer, that is queried and could cache data would help a  lot.  But I am not feeling willing to only return the "data necessary" for a given call because maybe UI changes happen or a new UI starts calling the RESTful api and expects or needs other data.

This is all in response to what I believe is called Linked Resources which is traditionally solved by HAL, JSON-LD, HATEOAS, etc..  However the traditional this is to return a link that can be followed to get the sub data.  But that devolves into a LOT of API requests back and forth. So the decision was made that instead we would do embeded objects. 

We did not go full HAL where a  `include=` can be specified to proactively include a resource instead we just have an on/off switch that includes everything or nothing.  (the more I read the more I wonder why I did not find anything about the HAL standard during my implementation... I don't remember seeing such a clean definition of the problem when I did my initial research) 

It will be a bit of a PITA to change this too much right now.  Quite a bit of the code using the APIs depend on this included information.  So what we really want to try to do is not include less, but instead find a way to include it faster and cleaner.

Reply:
> I think using a service, did i pick the right layer, that is queried and could cache data would help a lot. But I am not feeling willing to only return the "data necessary" for a given call because maybe UI changes happen or a new UI starts calling the RESTful api and expects or needs other data.

To continue on our earlier discussion, briefly mentioned, I think we're on the same page here. I think it's fine to return as much data as we can. I don't think we'd necessarily have to worry about caching just yet, we could come to that later if we truly need it. I hesitate because caching can make things really difficult to troubleshoot if we aren't careful. I'd think it's also good to consider API access patterns as well. APIs evolve over time, and it's difficult to know whether a piece of data you have now will necessarily be useful later, or if you will need to add something else or whether it would be easier for the server to handle it before it gets out of the API.

As an aside, I would make a case that in general, if we are approaching our APIs in an additive, backwards compatible way, it probably doesn't matter too much in the end if we've covered only the bare minimum number of bases, or if we need to add more to it. A service layer helps us to make those changes more quickly, and if we do it in a "attributes are added" manner, it also means we don't impact other API consumers. If UI Page A is using fields A, B, and C from some request, and we find we need to add D, E, and F to the API output for Page B, Page A remains unaffected because it isn't looking for fields D, E, and F.

That said, I'm not opposed to keeping things consistent and matching what is there now, as much as we can reasonably do so. Like you mentioned, we probably don't need the entire fallback hierarchy for a department, but we currently get it.

If our APIs are quick, hitting the API when needed probably isn't a big deal - for example, I think for the implementations you've described the idea is that you aren't necessarily making an immediate request for the next resource. I think there's a balance we could reach here as we get more and more of the UI converted to Vue, but that seems like a longer term discussion.

Also, we don't really leverage it well now, but we could eventually get things to a point in the UI where requests are made for additional resources when they appear on the screen, or perhaps find another strategy in the absolute worst case.

How do you feel about this, for this API call specifically (GET /departments)?

I can review our code for uses of this API and determine which attributes we currently reference from the API output. If I determine that something is missing from the output, I evaluate the amount of effort required to adjust on either the UI or the API. We may have some things we can add on the API side of things, but my initial feeling is that we can probably solve for it in the UI if needed.

Reply:
> > I think using a service, did i pick the right layer, that is queried and could cache data would help a lot. But I am not feeling willing to only return the "data necessary" for a given call because maybe UI changes happen or a new UI starts calling the RESTful api and expects or needs other data.
> 
> To continue on our earlier discussion, briefly mentioned, I think we're on the same page here. I think it's fine to return as much data as we can. I don't think we'd necessarily have to worry about caching just yet, we could come to that later if we truly need it. I hesitate because caching can make things really difficult to troubleshoot if we aren't careful. I'd think it's also good to consider API access patterns as well. APIs evolve over time, and it's difficult to know whether a piece of data you have now will necessarily be useful later, or if you will need to add something else or whether it would be easier for the server to handle it before it gets out of the API.
> 

I am thinking of a VERY limited sort of caching.  It could be helpful for a layer to cache replies done within a given request.  So with the above example If I am listing every division in a list and each division is looking down its fallback and so we get those LONG chains of fallbacks in slightly different orders repeated over and over.  However if the department service cached a given sql response for a given id, even just for that request, then we would greatly reduce our SQL queries to only once per division id.  That all by itself will save a LOT of sql work. 

> As an aside, I would make a case that in general, if we are approaching our APIs in an additive, backwards compatible way, it probably doesn't matter too much in the end if we've covered only the bare minimum number of bases, or if we need to add more to it. A service layer helps us to make those changes more quickly, and if we do it in a "attributes are added" manner, it also means we don't impact other API consumers. If UI Page A is using fields A, B, and C from some request, and we find we need to add D, E, and F to the API output for Page B, Page A remains unaffected because it isn't looking for fields D, E, and F.
> 
> That said, I'm not opposed to keeping things consistent and matching what is there now, as much as we can reasonably do so. Like you mentioned, we probably don't need the entire fallback hierarchy for a department, but we currently get it.

Yeah, I have a habit of trying to be too forward thinking when designing things like this and often it results in complexity that ends up being limited use.  I guess I am in a more willing to hack things up mood now.

> 
> If our APIs are quick, hitting the API when needed probably isn't a big deal - for example, I think for the implementations you've described the idea is that you aren't necessarily making an immediate request for the next resource. I think there's a balance we could reach here as we get more and more of the UI converted to Vue, but that seems like a longer term discussion.
> 

Yeah, the APIs being quick may be a request that would be hard to guarentee.  We just had that problem with logins where if too many people are logging in at the same time amazon was killing requests and people where left in bad states.  I do not know if increasing the number of http requests per action is going to be a good plan for us.

> Also, we don't really leverage it well now, but we could eventually get things to a point in the UI where requests are made for additional resources when they appear on the screen, or perhaps find another strategy in the absolute worst case.
> 
> How do you feel about this, for this API call specifically (GET /departments)?
> 
> I can review our code for uses of this API and determine which attributes we currently reference from the API output. If I determine that something is missing from the output, I evaluate the amount of effort required to adjust on either the UI or the API. We may have some things we can add on the API side of things, but my initial feeling is that we can probably solve for it in the UI if needed.

Ok lets focus down on just the `Department` resource and see.  `Staff` uses it so maybe trying to make `Staff` our first try for the new format was too far up the chain. 

`Cycle` would be even lower and simpler...  After `Cycle` could come `Event`.  But again I may be roadmapping out too far.

Reply:
> I am thinking of a VERY limited sort of caching. It could be helpful for a layer to cache replies done within a given request. So with the above example If I am listing every division in a list and each division is looking down its fallback and so we get those LONG chains of fallbacks in slightly different orders repeated over and over. However if the department service cached a given sql response for a given id, even just for that request, then we would greatly reduce our SQL queries to only once per division id. That all by itself will save a LOT of sql work.

So, the service layer works like this in essence. It fetches the necessary data for a given request and stitches things together or carries out business logic as needed. I don't think we would need to cache it within a given request, even if we did want to populate _all_ of the fallback department data.

> Yeah, I have a habit of trying to be too forward thinking when designing things like this and often it results in complexity that ends up being limited use. I guess I am in a more willing to hack things up mood now.

No worries. I do think there is value in trying to provide all of the stuff up front, and designing something that will help us make minimal changes to the APIs. When something evolves a bit too organically, it can result in some weird complexity on its own. There's definitely a benefit to finding a balance here.

> Yeah, the APIs being quick may be a request that would be hard to guarentee. We just had that problem with logins where if too many people are logging in at the same time amazon was killing requests and people where left in bad states. I do not know if increasing the number of http requests per action is going to be a good plan for us.

Agreed, but it also depends. I'm not suggesting we move to a strategy where we provide links to the next resource. We'd have some other things to work through before we could get there, such as handling requests a couple at a time rather than, for example, the 10 at once I was firing up on the ConCom List page. :)

> Ok lets focus down on just the Department resource and see.

OK, just to be sure - do you mean, _any call_ with Department data, or _only_ the List call?

Reply:
> > I am thinking of a VERY limited sort of caching. It could be helpful for a layer to cache replies done within a given request. So with the above example If I am listing every division in a list and each division is looking down its fallback and so we get those LONG chains of fallbacks in slightly different orders repeated over and over. However if the department service cached a given sql response for a given id, even just for that request, then we would greatly reduce our SQL queries to only once per division id. That all by itself will save a LOT of sql work.
> 
> So, the service layer works like this in essence. It fetches the necessary data for a given request and stitches things together or carries out business logic as needed. I don't think we would need to cache it within a given request, even if we did want to populate _all_ of the fallback department data.
> 

Perfect!

> > Yeah, I have a habit of trying to be too forward thinking when designing things like this and often it results in complexity that ends up being limited use. I guess I am in a more willing to hack things up mood now.
> 
> No worries. I do think there is value in trying to provide all of the stuff up front, and designing something that will help us make minimal changes to the APIs. When something evolves a bit too organically, it can result in some weird complexity on its own. There's definitely a benefit to finding a balance here.
> 
> > Yeah, the APIs being quick may be a request that would be hard to guarentee. We just had that problem with logins where if too many people are logging in at the same time amazon was killing requests and people where left in bad states. I do not know if increasing the number of http requests per action is going to be a good plan for us.
> 
> Agreed, but it also depends. I'm not suggesting we move to a strategy where we provide links to the next resource. We'd have some other things to work through before we could get there, such as handling requests a couple at a time rather than, for example, the 10 at once I was firing up on the ConCom List page. :)
> 
> > Ok lets focus down on just the Department resource and see.
> 
> OK, just to be sure - do you mean, _any call_ with Department data, or _only_ the List call?

Part of why I suggest `Cycle` or `Event` is that we could do the whole resource there instead of just call by call.  As a very first resource it may be easier to approach it that way.

Reply:
> Part of why I suggest Cycle or Event is that we could do the whole resource there instead of just call by call. As a very first resource it may be easier to approach it that way.

OK. `Event` is something that would be handled in the staff list call that comes after this PR. I'd much rather that we fix the performance issues first and then focus on the "best path forward".